0005-kzen-two-party-ecdsa.adoc

```diff
@@ -0,0 +1,173 @@
+
+The proof is described in <<Lindell17>> (section 6).
+It's a little tricky to get your head around.
+First P2 sends a challenge to P1.
+If P1's response to the challenge is valid then P2 knows that the encrypted private key corresponds to `p1_public_key`.
+
+The weird thing is that P1 doesn't actually use his knowledge of what private key is in order to produce the proof that the encrypted private key is correct.
+Anyone who knows the factorization `N = pq` can make this proof to P2.
+This is becasuse the proof is not showing knowledge of the private key, just that the encrypted private key is indeed the preimage of `p1_public_key`.
+
+==== Implementation notes
+
+This proof isn't in its own module but the code for it hangs around with the rest of the protocol:
+
+https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/party_one.rs
+(look for things starting with `PDL`).
+
+It looks like this protocol cannot be made non-interactive. It requires four rounds of communication.
+
+=== Range proof
+
+In order for the previous proof to actually prove the statement you have to couple it with a range proof which proves that the encrypted private key is in the curve order (i.e. is a valid private key).
+The poof chosen was originally from <<Boudot00>>  but I found it was easier to understand in <<Lindell17>> anyway (see Appendix A).
+
+The proof uses the cut and choose technique, so it's quite large.
+It's tricky to understand, but doesn't use any wonky math.
+You just have to follow what happens closely.
+
+==== Implementation notes
+
+To prove that the private key lies within the curve order P1 first has to choose their private key so that it's in `Z_q/3` rather than `Z_q`.
+Without this the proof will not be _complete_.
+
+It's implemented here:
+
+https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/range_proof.rs
+
+=== Messages
+
+Here's my early sketch of how many messages you need:
+
+1. P1 -> `Hash(p1_public_key, schnorr_nizk(p1_public_key))`
+2. P2 ->
+   1. `p2_public_key`
+   2. `schnorr_nizk(p2_public_key)`
+3. P1 ->
+   1. Opens commitment from first message
+   2. Paillier modulus `N`
+   3. Proof `N` was generated properly
+   4. `(c,r) = PaillierEncrypt(p1_private_key)`
+   5. Range proof for `c` encrypts a valid private key
+4. P2 -> Challenge for `c` being the private key of `p1_public_key`
+5. P1 -> Committed response for `c` being valid
+6. P2 -> Reveal challenge from (4)
+7. P1 -> Open response from (5)
+
+== Signing
+
+Assuming the keygen phase went well we have the parties knowing the following (where `x1`,`x2` are private keys and `X` is the joint public key).
+
+1. P1 knows: `x1` , `X`, `(N,p,q | N = pq)`,
+2. P2 knows: `x2`, `X`, `N`, `c1 = PaillierEncrypt(x1, N)`
+
+Now they want to sign a message `m`.
+
+Since ECDSA signatures are in the form `(r,s)`.
```

Comment:
:question: English question. Starting this sentence with `Since` makes me want to combine it with the 2nd one.
Right or wrong?

Reply:
Yes. A mistaken end of sentence.